src/main/java/io/confluent/connect/hdfs/DataWriter.java

```diff
@@ -451,11 +451,20 @@ public void close() {
     // data may have continued to be processed and we'd have to restart from the recovery stage,
     // make sure we apply the WAL, and only reuse the temp file if the starting offset is still
     // valid. For now, we prefer the simpler solution that may result in a bit of wasted effort.
-    for (TopicPartition tp : assignment) {
+    for (Map.Entry<TopicPartition, TopicPartitionWriter> ent : topicPartitionWriters.entrySet()) {
       try {
-        topicPartitionWriters.get(tp).close();
+        TopicPartitionWriter writer = ent.getValue();
+        if (writer != null) {
+          // In some failure modes, with might not have created a writer for all assignments
+          writer.close();
+        }
       } catch (ConnectException e) {
-        log.error("Error closing writer for {}. Error: {}", tp, e.getMessage());
+        log.warn(
+            "Unable to close writer for topic partition {}: {}",
+            ent.getKey(),
+            e.getMessage(),
+            e
+        );
```

Comment:
It's possible that the DataWriter is instantiated, the assignments set, but the TopicPartitionWriters are not created (again, due to an error). In this case, the close method attempts to fine the TPW for each assignment, and because there are none we get an NPE.

The new code iterates over any existing TPW we've created and closes them safely, ensuring to close all of them and log any errors. Because this is close, IMO we can log and don't need to propagate the exceptions.